### ISSUE TITLE
Tokens are acting strange recently

### ISSUE THREAD

**Issue by neico** [premake/premake-core#999]
Until recently this line worked fine: `location "%{_TARGET_OS}/%{_ACTION}"`  
but now it gives of `Error: unable to create directory '.../build|windows'`.

I wasn't able to pinpoint the code (change) where `|` is coming from yet, but the path is supposed to be `.../build/windows/vs2017`, suggesting that something is breaking off the translation entirely.

For now I'm using `location ( _TARGET_OS .. "/%{_ACTION}" )` instead which gives of the expected behaviour.

**Comment by neico** [comment]
I took another spin as this as it just keeps bothering me, and apparently it has a issue with having a token at the start of the string `location "%{_TARGET_OS}/%{_ACTION}"` or `"%{_ACTION}/%{_TARGET_OS}"` both end up with this strange vertical bar symbol, so it's definitely not limited to _TARGET_OS, insteringly tough, if you prepend ANY char before the percentage symbol it starts working:
`location " %{_TARGET_OS}/%{_ACTION}"`
`location "_%{_TARGET_OS}/%{_ACTION}"`
`location "%%{_TARGET_OS}/%{_ACTION}"`
`location "%%%{_TARGET_OS}/%{_ACTION}"`

I still fail to find the code section that might be at fault here, mostly due to the code jumping all over the place, especially around the detoken code.
I guess @starkos or @tvandijck might have more luck as they know the token parsing code better than myself or probably anyone else.

Something to note: it seems the same strings work fine with all api functions I tested it with where it think it could happen as well, but it seems location is pretty unique in that regard as it's one of the very few api functions that directly impact how premake interacts with the filesystem (at least related to solution and project file generation), and it's of `kind = "path"` with `tokens = true`, unlike others which are either `kind = "string"` (ex. `filename`) or have an additional `pathVars = true`

So maybe it's not something with the token parsing at all, but of how the resulting string is being read to build the filesystem paths?

**Comment by samsinsane** [comment]
The `|` character is likely coming from the deferred join system in the `path` code, I guess in [here](https://github.com/premake/premake-core/blob/master/src/host/path_join.c) somewhere the code is ignoring the first character in the path or something. Taking a glance at the code, I'm not entirely sure where the problem would be, but I'm sure stepping through the code with your example above will highlight the problem. I'm not sure when I'll have time to look into this further but if anyone else is able to look into it, that's probably where the issue is.

**Comment by neico** [comment]
I took a look at that code and I've got to agree that it probably is around there, following that I tried to trick the code by using `location "./%{_TARGET_OS}/%{_ACTION}"` but that resulted in the same error, so I guess it has a issue with parsing / ignoring tokens right after a path delimiter.

If I had a bit more time I'd actually step trough the code in-depth,  
but I currently don't have premake setup to be debuggable,  
so all I can do right now is fiddle with what I can do from inside a script and stare at the result in confusion...

Actually, after noticing that the join code is dependant on isabsolute I've found [this](https://github.com/premake/premake-core/blob/master/src/host/path_isabsolute.c#L71), which looks very suspiciously...

**Comment by samsinsane** [comment]
Had a closer look, it looks like the error actually occurs in `detoken.expand`, it will expand the tokens but only checks the token values if they're deferred paths (in `expandtoken`). I'm not sure if the value being passed in to `detoken.expand` should have `path.resolvedeferredjoin` called on it, or if it should be done within `detoken.expand`, perhaps within `recurse`. I manually added the following just before [this](https://github.com/premake/premake-core/blob/cda4f94e288393d6672c4e0d7fdbebee2d2439a3/src/base/detoken.lua#L190) function call:
```lua
if type(value) == "string" and path.hasdeferredjoin(value) then
  value = path.resolvedeferredjoin(value)
end
```
Before making that change, if you call `premake.detoken.expand(path.deferredjoin("path", "%{_TARGET_OS}/%{_ACTION}), {})` you'll see the result is something along the lines of `path|windows/vs2017`, and then after making the change it will be `path/windows/vs2017`. All of the tests continue to pass, but I'm a little hesitant to say that this is _the_ solution. I'll submit a PR tomorrow after I've cleaned everything else up.
